File: scripts/readme-i18n.mjs
import { createHash } from 'node:crypto';
import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import { resolve } from 'node:path';

const root = resolve(import.meta.dirname, '..');
const docs = resolve(root, 'docs/readme');
const languagesPath = resolve(docs, 'LANGUAGES.json');
const statePath = resolve(docs, 'TRANSLATION_STATE.json');
export const NAV_START = '<!-- readme-i18n:navigation:start -->';
export const NAV_END = '<!-- readme-i18n:navigation:end -->';

export function loadJson(path) { return JSON.parse(readFileSync(path, 'utf8')); }
export function languages() { return loadJson(languagesPath).languages; }
export function navigation(locale) {
  const parts = languages().map((item) => item.locale === locale
    ? `<strong>${item.name}</strong>`
    : `<a href="${item.file}">${item.name}</a>`);
  return `${NAV_START}\n<p align="center">${parts.join(' · ')}</p>\n${NAV_END}`;
}
export function replaceNavigation(text, locale) {
  const pattern = new RegExp(`${escapeRegExp(NAV_START)}[\\s\\S]*?${escapeRegExp(NAV_END)}`);
  if (pattern.test(text)) return text.replace(pattern, navigation(locale));
  const lines = text.split(/\r?\n/);
  let at = 1;
  while (at < lines.length && (!lines[at].trim() || lines[at].startsWith('[!'))) at += 1;
  lines.splice(at, 0, '', navigation(locale), '');
  return `${lines.join('\n').trimEnd()}\n`;
}
function escapeRegExp(text) { return text.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); }
export function semanticSourceText() {
  let text = readFileSync(resolve(root, 'README.md'), 'utf8');
  const navPattern = new RegExp(`${escapeRegExp(NAV_START)}[\\s\\S]*?${escapeRegExp(NAV_END)}`);
  text = text.replace(navPattern, '');
  text = text.replace(/^\[!\[.*?\]\(.*?\)\]\(.*?\)\s*$/gm, '');
  text = text.replace(/<!--\s*readme-section:[^>]+-->/g, '');
  text = text.replace(/<!--\s*readme-source-sha256:[^>]+-->/g, '');
  return text.replace(/\s+/g, ' ').trim();
}
export function sourceDigest() {
  return createHash('sha256').update(semanticSourceText(), 'utf8').digest('hex');
}
export function sourceSectionDigests() {
  const text = readFileSync(resolve(root, 'README.md'), 'utf8').replace(/\r\n/g, '\n');
  const ids = loadJson(resolve(docs, 'README_SCHEMA.json')).section_ids;
  const out = {};
  for (let i = 0; i < ids.length; i += 1) {
    const marker = `<!-- readme-section:${ids[i]} -->`;
    const start = text.indexOf(marker);
    if (start < 0) throw new Error(`Missing source section: ${ids[i]}`);
    const next = i + 1 < ids.length ? text.indexOf(`<!-- readme-section:${ids[i + 1]} -->`, start + marker.length) : text.length;
    const normalized = text.slice(start + marker.length, next < 0 ? text.length : next).replace(/\s+/g, ' ').trim();
    out[ids[i]] = createHash('sha256').update(normalized, 'utf8').digest('hex');
  }
  return out;
}
function loadState() {
  if (!existsSync(statePath)) return { schema: 'gpt-agent-manager.readme-translation-state.v1', source_locale: 'en', source_digest: '', translations: {} };
  return loadJson(statePath);
}
function writeState(state) { writeFileSync(statePath, `${JSON.stringify(state, null, 2)}\n`, 'utf8'); }
function syncNavigation() {
  for (const item of languages()) {
    const path = resolve(root, item.file);
    if (!existsSync(path)) continue;
    const text = readFileSync(path, 'utf8');
    writeFileSync(path, replaceNavigation(text, item.locale), 'utf8');
  }
  console.log(`README_I18N_NAV_SYNC languages=${languages().length}`);
}
function status() {
  const digest = sourceDigest();
  const sections = sourceSectionDigests();
  const state = loadState();
  console.log(`en\tSOURCE\t${digest.slice(0, 12)}`);
  for (const item of languages()) {
    if (item.locale === 'en') continue;
    const entry = state.translations?.[item.locale] ?? {};
    const staleSections = Object.keys(sections).filter((id) => entry.section_digests?.[id] !== sections[id]);
    const current = entry.status === 'current' && entry.source_digest === digest && staleSections.length === 0;
    const detail = current ? 'CURRENT' : `STALE: ${staleSections.join(',') || 'source'}`;
    console.log(`${item.locale}\t${detail}\ttier=${item.tier}`);
  }
}
function markCurrent(locales) {
  const allowed = new Set(languages().filter((x) => x.locale !== 'en').map((x) => x.locale));
  for (const locale of locales) if (!allowed.has(locale)) throw new Error(`Unknown locale: ${locale}`);
  const digest = sourceDigest();
  const sections = sourceSectionDigests();
  const state = loadState();
  state.source_digest = digest; state.source_sections = sections; state.translations ??= {};
  for (const locale of locales) state.translations[locale] = { source_digest: digest, status: 'current', section_digests: sections };
  writeState(state);
  console.log(`README_I18N_MARK_CURRENT locales=${locales.join(',')} source=${digest.slice(0, 12)}`);
}
function initState() {
  const digest = sourceDigest();
  const sections = sourceSectionDigests();
  const state = { schema: 'gpt-agent-manager.readme-translation-state.v1', source_locale: 'en', source_digest: digest, source_sections: sections, translations: {} };
  for (const item of languages()) if (item.locale !== 'en') state.translations[item.locale] = { source_digest: digest, status: 'current', section_digests: sections };
  writeState(state);
  console.log(`README_I18N_STATE_INIT languages=${languages().length} source=${digest.slice(0, 12)}`);
}
const [command, ...args] = process.argv.slice(2);
if (command === 'sync-navigation') syncNavigation();
else if (command === 'status') status();
else if (command === 'mark-current') markCurrent(args);
else if (command === 'init-state') initState();
else throw new Error('Usage: node scripts/readme-i18n.mjs <sync-navigation|status|mark-current|init-state> [locales...]');

File: scripts/readme_i18n.py
from __future__ import annotations

import argparse
import hashlib
import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DOCS = ROOT / "docs" / "readme"
LANGUAGES_PATH = DOCS / "LANGUAGES.json"
STATE_PATH = DOCS / "TRANSLATION_STATE.json"
NAV_START = "<!-- readme-i18n:navigation:start -->"
NAV_END = "<!-- readme-i18n:navigation:end -->"


def load_json(path: Path) -> dict:
    return json.loads(path.read_text(encoding="utf-8"))


def write_json(path: Path, value: dict) -> None:
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def languages() -> list[dict]:
    return load_json(LANGUAGES_PATH)["languages"]


def navigation(locale: str) -> str:
    parts: list[str] = []
    for item in languages():
        label = item["name"]
        if item["locale"] == locale:
            parts.append(f"<strong>{label}</strong>")
        else:
            parts.append(f'<a href="{item["file"]}">{label}</a>')
    return f'{NAV_START}\n<p align="center">{" · ".join(parts)}</p>\n{NAV_END}'


def replace_navigation(text: str, locale: str) -> str:
    block = navigation(locale)
    pattern = re.compile(re.escape(NAV_START) + r".*?" + re.escape(NAV_END), re.S)
    if pattern.search(text):
        return pattern.sub(block, text, count=1)
    lines = text.splitlines()
    insert_at = 1
    while insert_at < len(lines) and (not lines[insert_at].strip() or lines[insert_at].startswith("[!")):
        insert_at += 1
    lines[insert_at:insert_at] = ["", block, ""]
    return "\n".join(lines).rstrip() + "\n"


def semantic_source_text() -> str:
    text = (ROOT / "README.md").read_text(encoding="utf-8")
    text = re.sub(re.escape(NAV_START) + r".*?" + re.escape(NAV_END), "", text, flags=re.S)
    text = re.sub(r"^\[!\[.*?\]\(.*?\)\]\(.*?\)\s*$", "", text, flags=re.M)
    text = re.sub(r"<!--\s*readme-section:[^>]+-->", "", text)
    text = re.sub(r"<!--\s*readme-source-sha256:[^>]+-->", "", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


def source_digest() -> str:
    return hashlib.sha256(semantic_source_text().encode("utf-8")).hexdigest()


def load_state() -> dict:
    if not STATE_PATH.exists():
        return {"schema": "charterion.readme-translation-state.v1", "source_locale": "en", "source_digest": "", "translations": {}}
    return load_json(STATE_PATH)

def sync_navigation() -> None:
    for item in languages():
        path = ROOT / item["file"]
        if not path.exists():
            continue
        text = path.read_text(encoding="utf-8")
        path.write_text(replace_navigation(text, item["locale"]), encoding="utf-8")
    print(f"README_I18N_NAV_SYNC languages={len(languages())}")


def status() -> None:
    digest = source_digest()
    state = load_state()
    print(f"en\tSOURCE\t{digest[:12]}")
    translations = state.get("translations", {})
    for item in languages():
        locale = item["locale"]
        if locale == "en":
            continue
        entry = translations.get(locale, {})
        current = entry.get("source_digest") == digest and entry.get("status") == "current"
        state_label = "CURRENT" if current else "STALE"
        print(f"{locale}\t{state_label}\ttier={item['tier']}")


def mark_current(locales: list[str]) -> None:
    available = {item["locale"] for item in languages() if item["locale"] != "en"}
    unknown = sorted(set(locales) - available)
    if unknown:
        raise SystemExit(f"Unknown locale(s): {', '.join(unknown)}")
    digest = source_digest()
    state = load_state()
    state["source_digest"] = digest
    translations = state.setdefault("translations", {})
    for locale in locales:
        translations[locale] = {"source_digest": digest, "status": "current"}
    write_json(STATE_PATH, state)
    print(f"README_I18N_MARK_CURRENT locales={','.join(locales)} source={digest[:12]}")


def init_state() -> None:
    digest = source_digest()
    state = {
        "schema": "charterion.readme-translation-state.v1",
        "source_locale": "en",
        "source_digest": digest,
        "translations": {},
    }
    for item in languages():
        locale = item["locale"]
        if locale == "en":
            continue
        state["translations"][locale] = {"source_digest": digest, "status": "current"}
    write_json(STATE_PATH, state)
    print(f"README_I18N_STATE_INIT languages={len(languages())} source={digest[:12]}")


def main() -> None:
    parser = argparse.ArgumentParser(description="Manage multilingual README metadata and navigation.")
    sub = parser.add_subparsers(dest="command", required=True)
    sub.add_parser("sync-navigation")
    sub.add_parser("status")
    sub.add_parser("init-state")
    mark = sub.add_parser("mark-current")
    mark.add_argument("locales", nargs="+")
    args = parser.parse_args()
    if args.command == "sync-navigation":
        sync_navigation()
    elif args.command == "status":
        status()
    elif args.command == "init-state":
        init_state()
    elif args.command == "mark-current":
        mark_current(args.locales)


if __name__ == "__main__":
    main()
